### DISCUSSION TITLE
CI/CD de app em NodeJS, como fazem?

### DISCUSSION THREAD

**Discussion by wilsonneto-swe** [frontendbr/forum#1583] [Perguntas e Respostas]
Galera, estou com dúvidas referentes ao deploy de uma app nodejs, ao fazer o deploy...

Estou rodando a aplicação com "node server.js" e deixo esta janela aberta no servidor escutando a porta 3333, um nginx servido na porta 80 os arquivos estáticos e direcionando as requisições de api para o node, isso em um debian.

Mas meu problema é, meu pipeline de deploy é disparado no meu commit na master, ele envia os arquivos por ftp para este servidor e sobrescreve os arquivos ali na pasta do projeto, mas ai eu tenho que matar o processo do "node server.js" e iniciar de novo? 
Ou posso usar o nodemon em produção para se atualizar automaticamente?
Como fazer para que um processo fique vigiando a aplicação e se ela cair voltar a subir ela?

Meu primeiro deploy de app em node em um servidor de produção... por isso tantas dúvidas, meu server é um vps debian, eu estou fazendo manualmente, colo os arquivos e executo o que for necessário, mas quero passar tudo isso para um pipeline de deploy automático.

Muito obrigado, 
abs!

**Comment by kavalcante** [comment]
> Como fazer para que um processo fique vigiando a aplicação e se ela cair voltar a subir ela?

PM2

Utilizamos o [Buddy](https://buddy.works) para rodar os testes, buildar, dar npm install e depois restartar o PM2 (esse tem o comando de npm start com as .env de produção).

**Comment by wilsonneto-swe** [comment]
Show @kavalcante  muito obrigado, me ajudou muito essa dica!

**Comment by wilsonneto-swe** [comment]
Show, mas como seria o fluxo dos jobs do pipeline que usam @kavalcante? 
[... build ...]
[ subir os arquivos buildados para o servidor via ftp ]
[ executar um comando no servidor para matar o processo atual do node para assim ele recomeçar pegando já os novos arquivos ]

Seria isto?

Muito obrigado, abraços!
